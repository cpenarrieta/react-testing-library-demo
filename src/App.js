import React, { useState } from "react";
import logo from "./logo.svg";
import "./App.css";

function App() {
  const [count, setCount] = useState(0);

  const increment = () => setCount(count + 1);
  const decrement = () => setCount(count - 1);
  const incrementAsync = () => setTimeout(() => setCount(count + 1), 100);
  const decrementAsync = () => setTimeout(() => setCount(count - 1), 100);

  return (
    <div className="App">
      <header className="App-header">
        <h1>Learning react-testing-library</h1>
        <img src={logo} className="App-logo" alt="logo" />
        <p data-testid="counter-test-id">Current counter: {count}</p>
        <div className="flex-row">
          <button className="main-button" onClick={increment}>
            +
          </button>
          <button className="main-button" onClick={decrement}>
            -
          </button>
        </div>
        <div className="flex-row">
          <button className="main-button" onClick={incrementAsync}>
            + Async
          </button>
          <button className="main-button" onClick={decrementAsync}>
            - Async
          </button>
        </div>
      </header>
    </div>
  );
}

export default App;
