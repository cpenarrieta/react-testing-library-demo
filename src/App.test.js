import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, fireEvent, waitForElement } from "@testing-library/react";
import App from "./App";

it("renders without crashing", () => {
  const { getByText, getByTestId, asFragment } = render(<App />);

  expect(asFragment()).toMatchSnapshot();
  expect(getByText("Learning react-testing-library")).toBeInTheDocument();
  expect(getByTestId("counter-test-id")).toHaveTextContent("0");
});

// Fire Events

it("increment counter correctly", () => {
  const { getByText, getByTestId } = render(<App />);

  fireEvent.click(getByText("+"));

  expect(getByTestId("counter-test-id")).toHaveTextContent("1");
});

// fireEvent can also fill form and much more
// fireEvent.change(getByLabelText(/username/i), {target: {value: 'chuck'}})
// fireEvent.change(getByLabelText(/password/i), {target: {value: 'norris'}})

it("decrease counter correctly", () => {
  const { getByText, getByTestId } = render(<App />);

  fireEvent.click(getByText("-"));

  expect(getByTestId("counter-test-id")).toHaveTextContent("-1");
});

it("increment counter multiple times", () => {
  const { getByText, getByTestId } = render(<App />);

  const plusButton = getByText("+");
  fireEvent.click(plusButton);
  fireEvent.click(plusButton);
  fireEvent.click(plusButton);

  expect(getByTestId("counter-test-id")).toHaveTextContent("3");
});

it("increment/decrement counter multiple times", () => {
  const { getByText, getByTestId } = render(<App />);

  const plusButton = getByText("+");
  const minusButton = getByText("-");
  fireEvent.click(plusButton);
  fireEvent.click(minusButton);
  fireEvent.click(plusButton);
  fireEvent.click(plusButton);
  fireEvent.click(minusButton);

  expect(getByTestId("counter-test-id")).toHaveTextContent("1");
});

// Async Events

// this test fails
xit("increment async counter correctly - wrong test", () => {
  const { getByText, getByTestId } = render(<App />);

  fireEvent.click(getByText("+ Async"));

  expect(getByTestId("counter-test-id")).toHaveTextContent("1");
});

it("increment async counter correctly", async () => {
  const { getByText } = render(<App />);

  fireEvent.click(getByText("+ Async"));

  await waitForElement(() => getByText("Current counter: 1"));
});

it("decrement async counter correctly", async () => {
  const { getByText } = render(<App />);

  fireEvent.click(getByText("- Async"));

  await waitForElement(() => getByText("Current counter: -1"));
});
