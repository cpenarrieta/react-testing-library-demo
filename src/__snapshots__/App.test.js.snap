// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <h1>
        Learning react-testing-library
      </h1>
      <img
        alt="logo"
        class="App-logo"
        src="logo.svg"
      />
      <p
        data-testid="counter-test-id"
      >
        Current counter: 0
      </p>
      <div
        class="flex-row"
      >
        <button
          class="main-button"
        >
          +
        </button>
        <button
          class="main-button"
        >
          -
        </button>
      </div>
      <div
        class="flex-row"
      >
        <button
          class="main-button"
        >
          + Async
        </button>
        <button
          class="main-button"
        >
          - Async
        </button>
      </div>
    </header>
  </div>
</DocumentFragment>
`;
